### PR TITLE
Expose getting/setting 'authority' field in Python

### LINKED ISSUE #12400: Cannot override the default HTTP/2 authority from Python when creating call objects
gRPC 1.7.0-dev, Python 2.7, Linux 4.4 (Ubuntu 14.04)

Once a channel has been created, there is no way to specify an alternative value for the `:authority` HTTP/2 pseudo-header from Python. gRPC categorically rejects any metadatum that specifies an HTTP/2 pseudo-header. In addition, the Python bindings do not provide any way to supply an argument to the `grpc_channel_create_call()` function's authority parameter. As a result, Python users are thus stuck with whatever value was set from the `GRPC_ARG_DEFAULT_AUTHORITY` channel argument. (This is especially irksome using the `grpc_insecure_channel_create_from_fd()` function, which for some reason hardcodes the authority to `test.authority`. I'm aware that no official Python bindings exist for that function; I have some of my own that I'd eventually like to upstream.)

### PR DESCRIPTION
**Pull request by kpayson64**
Fixes #12400

### CODE REVIEW

File: src/python/grpcio_tests/tests/unit/_host_test.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+# Copyright 2016 gRPC authors.
```
Comment (mehrdada) [inline src/python/grpcio_tests/tests/unit/_host_test.py]: Nit: s/2016/2018

File: src/python/grpcio/grpc/_channel.py [hunk 2]
```diff
@@ -464,15 +464,16 @@ def _prepare(self, request, timeout, metadata):
...
+    def _blocking(self, request, timeout, metadata, credentials, host):
         state, operations, deadline, rendezvous = self._prepare(
             request, timeout, metadata)
         if rendezvous:
             raise rendezvous
         else:
             completion_queue = cygrpc.CompletionQueue()
-            call = self._channel.create_call(None, 0, completion_queue,
-                                             self._method, None, deadline)
+            encoded_host = _common.encode(host) if host is not None else None
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_channel.py]: Flip the branches and drop the `not`.

File: src/python/grpcio_tests/tests/unit/_host_test.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class _MethodHandler(grpc.RpcMethodHandler):
+
+    def __init__(self):
+        self.request_streaming = False
+        self.response_streaming = False
+        self.request_deserializer = None
+        self.response_serializer = None
+        self._host = None
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/unit/_host_test.py]: We expect this to be accessed from different threads; protect it with a `threading.Condition`.

File: src/python/grpcio_tests/tests/unit/_host_test.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    def stream_stream(self, request_iterator, servicer_context):
+        self._host = servicer_context.host()
+        # TODO(issue:#6891) We should be able to remove this loop,
+        # and replace with return; yield
+        for request in request_iterator:
+            yield _RESPONSE
+
+    def host(self):
+        return self._host
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/unit/_host_test.py]: Use the `threading.Condition` to block here until this is non-`None`?

File: src/python/grpcio_tests/tests/unit/_host_test.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class _GenericHandler(grpc.GenericRpcHandler):
+
+    def __init__(self):
+        self._handler = _MethodHandler()
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/unit/_host_test.py]: `self._host_cell = [None]`? Then in `service` you can just have `return _MethodHandler(self._host_cell, <False or True>, <False or True>)`?

File: src/python/grpcio_tests/tests/unit/_host_test.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    def setUp(self):
+        self._server = test_common.test_server()
+        self._handler = _GenericHandler()
+        self._server.add_generic_rpc_handlers((self._handler,))
+        self._port = self._server.add_insecure_port('[::]:0')
+        self._server.start()
+        self._channel = grpc.insecure_channel('localhost:%d' % self._port)
+
+    def tearDown(self):
+        self._server.stop(0)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio_tests/tests/unit/_host_test.py]: Pass `None` here? If the server fails to stop we'd like the test itself to fail, right?

**Comment by mehrdada** [comment]
Do we need to wait for this to land before branch cut? If yes please tag it as such

**Comment by kpayson64** [comment]
@mehrdada Can you point to exactly where this should be added to the interceptor examples?  I didn't see a use case for and this should be a non-breaking API change.

**Comment by mehrdada** [comment]
@kpayson64 specifically classes that implement the ClientCallDetails interface like https://github.com/grpc/grpc/blob/master/examples/python/interceptors/headers/header_manipulator_client_interceptor.py#L22

**Comment by kpayson64** [comment]
@mehrdada 
I can make that small change, but I'm actually pretty worried about that.

AFAICT, not updating the example will cause it to break?  That would make this a breaking api change right?

**Comment by mehrdada** [comment]
#14105 has been merged, unblocking this.

**Comment by nathanielmanistaatgoogle** [comment]
This needs conflict resolution and its list of commits currently looks completely crazypants. Also I think it still has outstanding code review comments?

**Comment by kpayson64** [comment]
@nathanielmanistaatgoogle 
I'm calling it host for consistency with the gRPC core.  I saw another comment in bug thread about how this can be represented differently on different transports.

@mehrdada 
I've updated the example.

**Comment by nathanielmanistaatgoogle** [comment]
With apologies for being a stick in the mud: this is our application-facing API; we need to build it on a stronger foundation than mere consistency with gRPC Core.

Does setting or not setting this string on the invocation side of an RPC have an effect on gRPC behavior other than the string being present or absent on the service side of the RPC? If so, what? If not, why is it an API feature rather than something to be set (or not set) in the user-space that is the RPC's metadata?

If it is sometimes called "host", does that correspond to [any other definition of "host" in networking](https://en.wikipedia.org/wiki/Host_(network))? If it is sometimes called "authority", does that correspond to any other definition of "authority" in information security? Can we call it "hiybbprqag" without contradicting anything else we've said about it?

Thinking beyond gRPC Core: do other gRPC languages present this feature in their application-facing APIs? Which? What do they say to their users about it?

Regarding "I saw another comment in bug thread about how this can be represented differently on different transports": is this a feature by which we expose certain underlying HTTP2 concepts or implementation to applications? Doing so it not necessarily wrong but it is something about which we should be _even more careful_ (or fussy) than otherwise.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++

 -------------- SHRINK --------------
  [ = ]       0 [None]     -16  -0.0%

  [ = ]       0 TOTAL      -16  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: src/core/lib/security/transport/client_auth_filter.cc [hunk 7]
```diff
@@ -429,3 +430,8 @@ const grpc_channel_filter grpc_client_auth_filter = {
     destroy_channel_elem,
     grpc_channel_next_get_info,
     "client-auth"};
+
```
Comment (dgquintas) [inline src/core/lib/security/transport/client_auth_filter.cc]: please remove all these new empty lines

**Comment by dgquintas** [comment]
clang-format isn't happy

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_fullstack_streaming_pump.BM_PumpStreamServerToClient_SockPair__512.opt.new: 1


[microbenchmarks] No significant performance differences
```

File: src/core/lib/security/transport/client_auth_filter.cc (hunk 7 above)
Comment (kpayson64) [reply src/core/lib/security/transport/client_auth_filter.cc]: Sorry, pushed a commit to the wrong PR (and OWNERS added you)

**Comment by kpayson64** [comment]
@nathanielmanistaatgoogle 
I've renamed host to authority.  PTAL

**Comment by nathanielmanistaatgoogle** [comment]
Yes, but I'd still like to learn a lot more about this new feature in the application-exposed API. How does setting it affect the client-side gRPC objects? How does setting it affect the server-side gRPC objects? How, and for what purpose, are applications intended to use it? Are there any ways that applications may use it that despite compiling and executing are unsupported? Why does it need to exist as a concept in the gRPC Python API - why can't it just be something set in the metadata by the application on one side and extracted from the metadata by the application on the other side? (Don't answer here; all this context belongs in the commit log message.)

**Comment by mehrdada** [comment]
There is an issue filed recently (https://github.com/grpc/grpc/issues/14420) that is about getting access to the peer TLS public key. I was wondering if instead of passing host, it would be possible to pass an arbitrary function to verify if host is valid, i.e. matches a certain pattern, etc.?

**Comment by kpayson64** [comment]
I don't have all the answers to when/why this field should be set.  I suspect that is something that is language agnostic and should be documented grpc-wide.

In both c-core and Java, this is an out-of-band metadata field.  I suspect it is because it is defined in the HTTP2 RFC as a pseudo header (and given a default value).  If we are going to buck the trend and have this treated like normal metadata, we should have a really strong reason for doing so.

@mehrdada 
That issue seems out of the scope of this PR.

**Comment by mwisniewski0** [comment]
> I was wondering if instead of passing host, it would be possible to pass an arbitrary function to verify if host is valid, i.e. matches a certain pattern, etc.?

@mehrdada This feature would be even better than the TLS raw public key auth. Raw public key auth could be very easily implemented as a predicate, and other custom verification options would be available. Should I edit my request/open a new issue?

**Pull request closed without merging**